### PR TITLE
[Feature] Support various scales in RRDBNet

### PR DESCRIPTION
**Pull request by anse3832**
## Motivation
1. Current RRDBNet only supports 4x upsampling.
2. The parameter "radius" in config file of real_esrgan is wrong.

## Modification
- Add scale parameter in rrdb_net
- Edit config file of real_esrgan for supporting various scales
- Edit config file of real_esrgan for fixing some errors (radius=50 -> kernel_size=51)

### CODE REVIEW

File: mmedit/models/backbones/sr_backbones/rrdb_net.py [hunk 1]
```diff
@@ -9,6 +9,25 @@
...
+    """
+    b, c, hh, hw = x.size()
+    out_channel = c * (scale**2)
+    assert hh % scale == 0 and hw % scale == 0
+    h = hh // scale
+    w = hw // scale
+    x_view = x.view(b, c, h, scale, w, scale)
+    return x_view.permute(0, 1, 3, 5, 2, 4).reshape(b, out_channel, h, w)
+
+
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/rrdb_net.py:30]: You can import this function using the following command: 

```
from mmedit.models.common import pixel_unshuffle
```

File: mmedit/models/backbones/sr_backbones/rrdb_net.py [hunk 2]
```diff
@@ -111,10 +130,19 @@ class RRDBNet(nn.Module):
...
                  mid_channels=64,
                  num_blocks=23,
                  growth_channels=32):
         super().__init__()
+        assert scale in [1, 2,
+                         4], 'Current code only supports scales of 1, 2, 4'
+        if scale == 2:
+            in_channels = in_channels * 4
+        elif scale == 1:
+            in_channels = in_channels * 16
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/rrdb_net.py:143]: Suggest to rewrite it as:
```
if scale == 2:
    in_channels = in_channels * 4
elif scale == 1:
    in_channels = in_channels * 16
elif scale != 4:
    raise ValueError('Currently support only (1x, 2x, 4x) upsampling only.')
```

File: mmedit/models/backbones/sr_backbones/rrdb_net.py [hunk 3]
```diff
@@ -139,15 +167,23 @@ def forward(self, x):
         Returns:
             Tensor: Forward results.
         """
+        if self.scale == 2:
+            feat = pixel_unshuffle(x, scale=2)
+        elif self.scale == 1:
+            feat = pixel_unshuffle(x, scale=4)
+        else:
+            feat = x
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/sr_backbones/rrdb_net.py:175]:
```
if self.scale in [1, 2]:
    feat = pixel_unshuffle(x, scale=4 // self.scale)
else:
    feat = x
```

**Review by ckkelvinchan — COMMENTED** [review]
@anse3832 Thank you very much! Would you mind changing also [realesrnet_c64b23g32_12x4_lr2e-4_1000k_df2k_ost.py](https://github.com/open-mmlab/mmediting/blob/master/configs/restorers/real_esrgan/realesrnet_c64b23g32_12x4_lr2e-4_1000k_df2k_ost.py) and adding the unittest for the added support [here](https://github.com/open-mmlab/mmediting/blob/23213c839ff2d1907a80d6ea29f13c32a24bb8ef/tests/test_models/test_backbones/test_sr_backbones/test_sr_backbones.py#L145)?

Thanks again!

File: tests/test_models/test_backbones/test_sr_backbones/test_sr_backbones.py [hunk 4]
```diff
@@ -164,6 +165,26 @@ def test_rrdbnet_backbone():
...
+    img = _demo_inputs(input_shape)
+    output = net(img)
+    assert output.shape == (1, 3, 24, 24)
+
+    # x2 model forward (gpu)
+    if torch.cuda.is_avaliable():
+        net = net.cuda()
+        output = net(img.cuda())
+        assert output.shape == (1, 3, 24, 24)
+
```
Comment (ckkelvinchan) [inline tests/test_models/test_backbones/test_sr_backbones/test_sr_backbones.py:187]: The unittest does not pass. Here you can simply copy the original one and set `scale` to 2.

**Comment by ckkelvinchan** [comment]
Hi @anse3832, it seems that linting failed. You can first install pre-commit by `pip install pre-commit`. Then you can type `pre-commit install` in the root directory of mmediting. 

After that every time you commit something, it will check the format for you. Thank you very much for your contributions and we are almost there!

**Comment by ckkelvinchan** [comment]
In addition, unittest also failed. You may try to run the test locally to test whether it has any errors~

**Comment by nijkah** [comment]
I'm a member of the team with anse3832.
I helped his commit's lint & pytest.
Can you check it?

**Comment by ckkelvinchan** [comment]
> I'm a member of the team with anse3832. I helped his PR, lint and pytest. Can you check it?

Hi @nijkah, thank you very much! I will check it.

**Comment by ckkelvinchan** [comment]
Hi @anse3832 @nijkah, I think it is almost okay. Could you help adding one more unittest, where scale=3? We want to cover the case when scale is not 1,2 or 4

**Comment by anse3832** [comment]
@ckkelvinchan
I have some concerns about ESRGAN and Real-ESRGAN

1. The Real-ESRGAN official code (https://github.com/xinntao/Real-ESRGAN) only works when scale is 1,2 or 4. Therefore, we I'm worried about supporting x3 scale.

2. The ESRGAN and Real-ESRGAN use the same "RRDBNet", but it works slightly different.
 - Real-ESRGAN uses pixelunshuffle, while ESRGAN doesn't
 - Real-ESRGAN always uses nearest upsample twice regardless of scale (1, 2, 4), while ESRGAN use different times of nearest upsample according to scale (once in x2, x3 scale and twice in x4 scale)
 
Considering above concerns, I suggest newly writing the code for "RRDBNet" in Real-ESRGAN as "RealRRDBNet"
 - When user wants to use ESRGAN, the code calls "RRDBNet" (it supports x2, x3 and x4 scale)
 - When user wants to use Real-ESRGAN, the code calls "RealRRDBNet" (it supports x1, x2 and x4 scale) 
 Is it okay to proceed these works? If you have any suggestions, please let me know.

**Comment by ckkelvinchan** [comment]
Thank you for your suggestions! What I meant in the previous reply is that we want to cover the following line of code.

image

That means we are not going to support x3. We just need to make sure in the unittest that an error will be raised. I checked the official code and it seems that x3 is not supported neither.

**Comment by anse3832** [comment]
@ckkelvinchan
Do you mean that the unittest code should be like below? (it will raise error because of x3)
error

**Comment by ckkelvinchan** [comment]
> @ckkelvinchan Do you mean that the unittest code should be like below? (it will raise error because of x3) error

When you expect an error, you should use `with pytest.raises()`. You can follow the code [here](https://github.com/open-mmlab/mmediting/blob/23213c839ff2d1907a80d6ea29f13c32a24bb8ef/tests/test_models/test_backbones/test_sr_backbones/test_glean_net.py#L43). 

By the way, you do not need to separate RealRRDBNet from RRDBNet. That means, we support only 1x, 2x, and 4x only, and for 1x and 2x, we use an additional pixel-unshuffle.

Thanks again!

**Comment by anse3832** [comment]
@ckkelvinchan 
In the official ESRGAN code (https://github.com/xinntao/ESRGAN/blob/master/RRDBNet_arch.py), it only supports x4 scale and it does not use pixelunshuffle.

If it is allowed to use pixelunshuffle in ESRGAN, I did not need to separate the code for ESRGAN and Real-ESRGAN. In this case, the code supports for x1, x2 or x4 scale. (Neither of official ESRGAN or Real-ESRGAN supports x3 scale)
- RRDBNet (both in ESRGAN and Real-SRGAN): supports x1, x2 or x4 scale and it uses pixelunshuffle

If I should follow the official code, I think it's better to separate the code.
- RRDBNet (in ESRGAN): supports x4 scale and it doesn't use pixelunshuffle
- RealRRDBNet (in Real-ESRGAN): supports x1, x2 or x4 scale and it uses pixelunshuffle

**Comment by ckkelvinchan** [comment]
> 

Yes it is allowed to use pixel-unshuffle in ESRGAN. Since this modification does not affect the original ESRGAN. We can extend it to support x1 and x2.

**Review by ckkelvinchan — DISMISSED** [review]
Thank you @anse3832 and @nijkah !

**Comment by ckkelvinchan** [comment]
Hi @anse3832 it seems that CI failed. Could you help have a look? Thanks again for your help.

**Comment by nijkah** [comment]
@ckkelvinchan I think it works this time. 🙏

**Comment by nijkah** [comment]
Check lint failure.
@anse3832  Can you approve the PR?
https://github.com/anse3832/mmediting/pull/4

**Review by ckkelvinchan — APPROVED** [review]
Thanks @anse3832 and @nijkah ! All tests passed.

**Comment by ckkelvinchan** [comment]
Your efforts are very much appreciated. Look forward to working with you next time!

**Comment by wangruohui** [comment]
Hi all, it seems that this PR accidentally changes file permissions of some config files from 644 to 755, can anybody help to fix that?

**Comment by nijkah** [comment]
@wangruohui 
https://github.com/nijkah/mmediting/commit/0d41d2cea9873c9a142f63f9e89573bda77ec1bc
Is it okay to pull request this branch?

**Comment by ckkelvinchan** [comment]
> @wangruohui
> [nijkah@0d41d2c](https://github.com/nijkah/mmediting/commit/0d41d2cea9873c9a142f63f9e89573bda77ec1bc)
> Is it okay to pull request this branch?

It is fixed in #718 by @wangruohui. Thanks anyways!

**Comment by OpenMMLab-Assistant003** [comment]
Hi @anse3832！First of all, we want to express our gratitude for your significant PR in the MMEditing project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤ @anse3832

**Pull request merged**